jlm/rvsdg/NodeNormalization.hpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2024 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_RVSDG_NODENORMALIZATION_HPP
+#define JLM_RVSDG_NODENORMALIZATION_HPP
+
+#include <vector>
+
+namespace jlm::rvsdg
+{
+
+class output;
+
+/**
+ * \brief Normalizes the operands of an operation.
+ *
+ * The usage expectation is that IsApplicable() is invoked to determine whether operands can be
+ * normalized BEFORE ApplyNormalization() is invoked.
+ *
+ * @tparam TOperation The operation for which to normalize the operands.
+ */
+template<class TOperation>
+class NodeNormalization
+{
+  static_assert(
+      std::is_base_of<Operation, TOperation>::value,
+      "Template parameter TOperation must be derived from jlm::rvsdg::operation.");
+
+public:
+  virtual ~NodeNormalization() noexcept = default;
+
+  NodeNormalization() = default;
+
```

Comment:
given that this is just an interface, why explicit about all of copy & assign? that I think belongs to implementation

Reply:
Can you maybe also explain in what places that interface would be used? I think having a base interface is only useful when you somehow abstract over the actual implementations (i.e. require the interface in a certain place without knowing the underlying implementation). In all places that it is wired up to in Load.cpp actually you never need an interface, free functions would do as well.

Reply:
I can push the copy & assign down to the leave classes.

Regarding the interface: See comment above.

Reply:
I move copy & paste constructors/operators to leave classes